drivers/sensor/mb7040/mb7040.c

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright (c) 2025 Sabrina Simkhovich <sabrinasimkhovich@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT maxbotix_mb7040
+
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/device.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/drivers/gpio.h>
+
+#define RANGE_CMD              0x51
+#define MB7040_HAS_STATUS_GPIO DT_ANY_INST_HAS_PROP_STATUS_OKAY(status_gpios)
+
+LOG_MODULE_REGISTER(mb7040, LOG_LEVEL_DBG);
+
+struct mb7040_data {
+	uint16_t distance_cm;
+	struct k_sem read_sem;
+#if MB7040_HAS_STATUS_GPIO
+	struct gpio_callback gpio_cb;
+#endif
+};
+
+struct mb7040_config {
+	struct i2c_dt_spec i2c;
+#if MB7040_HAS_STATUS_GPIO
+	struct gpio_dt_spec status_gpio;
+#endif
+};
+
+#if MB7040_HAS_STATUS_GPIO
+static void status_gpio_callback(const struct device *dev, struct gpio_callback *cb, uint32_t pins)
+{
+	struct mb7040_data *data = CONTAINER_OF(cb, struct mb7040_data, gpio_cb);
+	k_sem_give(&data->read_sem);
+}
+#endif
+
+static int mb7040_sample_fetch(const struct device *dev, enum sensor_channel chan)
+{
+	const struct mb7040_config *cfg = (struct mb7040_config *)dev->config;
+	struct mb7040_data *data = (struct mb7040_data *)dev->data;
+	uint8_t cmd = RANGE_CMD;
+	int ret;
+	uint8_t read_data[2];
+
+	if (chan != SENSOR_CHAN_DISTANCE && chan != SENSOR_CHAN_ALL) {
+		LOG_ERR("Sensor only supports distance");
+		return -EINVAL;
+	}
+
+	k_sem_reset(&data->read_sem);
+
+#if MB7040_HAS_STATUS_GPIO
+	/* Enable interrupt before writing */
+	ret = gpio_pin_interrupt_configure_dt(&cfg->status_gpio, GPIO_INT_EDGE_FALLING);
+	if (ret != 0) {
+		LOG_ERR("Failed to configure interrupt: %d", ret);
+		return ret;
+	}
+#endif
+
+	/* Write range command to sensor */
+	ret = i2c_write_dt(&cfg->i2c, &cmd, 1);
+	if (ret != 0) {
+		LOG_ERR("I2C write failed with error %d", ret);
+		return ret;
+	}
+
+	ret = k_sem_take(&data->read_sem, K_MSEC(CONFIG_MB7040_DELAY_MS));
+	if (ret != 0) {
+		if (ret == -EAGAIN) {
+			LOG_ERR("Semaphore take timed out");
+			return -ETIMEDOUT;
```

Comment:
Logic looks incomplete to me.

When `status_gpio` is present:
* 0 - success
* `-EAGAIN` - timeout
* other - some error

When `status_gpio` is missing:
* 0 - some internal error (there is no handler to set it except for the gpio interrupt handler which is not used)
* `-EAGAIN` - success
* other - some error